File: packages/ts/cli/src/flags.ts
/**
 * Argv parsing + global-flag resolution. Precedence for resolved values is
 * explicit flag > env var > (left to the request/config layer) default — the
 * same precedence the Python twin applies.
 */

/** The shape of `argv` after a single pass. */
export interface ParsedArgv {
  command?: string;
  args: string[];
  flags: Record<string, string | number | boolean | undefined>;
  help: boolean;
}

const NUMERIC_FLAGS = new Set(["temperature", "max-tokens"]);

/**
 * One-pass argv parser: first bare token is the command, the rest are positional
 * args; `--key value` / `--key=value` become flags; `--help`/`-h` set `help`.
 * Numeric flags (`--temperature`, `--max-tokens`) are coerced to numbers.
 */
export function parseArgv(argv: string[]): ParsedArgv {
  const parsed: ParsedArgv = { args: [], flags: {}, help: false };
  for (let i = 0; i < argv.length; i++) {
    const token = argv[i] as string;
    if (token === "--help" || token === "-h") {
      parsed.help = true;
      continue;
    }
    if (token.startsWith("--")) {
      const body = token.slice(2);
      const eq = body.indexOf("=");
      let key: string;
      let value: string | boolean;
      if (eq >= 0) {
        key = body.slice(0, eq);
        value = body.slice(eq + 1);
      } else {
        key = body;
        const next = argv[i + 1];
        if (next !== undefined && !next.startsWith("--")) {
          value = next;
          i++;
        } else {
          value = true;
        }
      }
      parsed.flags[key] =
        typeof value === "string" && NUMERIC_FLAGS.has(key) ? Number(value) : value;
      continue;
    }
    if (parsed.command === undefined) parsed.command = token;
    else parsed.args.push(token);
  }
  return parsed;
}

/** The four global flags after env fallback, in a strongly-typed shape. */
export interface ResolvedFlags {
  provider?: string;
  model?: string;
  temperature?: number;
  maxTokens?: number;
}

/**
 * Resolve the global flags with precedence explicit flag > env. `--provider`
 * falls back to `LLMORCH_PROVIDER`; the rest have no env layer of their own.
 */
export function resolveFlags(
  flags: Record<string, string | number | boolean | undefined>,
  env: NodeJS.ProcessEnv,
): ResolvedFlags {
  const num = (v: unknown): number | undefined =>
    typeof v === "number" ? v : typeof v === "string" ? Number(v) : undefined;
  return {
    provider: (flags.provider as string | undefined) ?? env.LLMORCH_PROVIDER,
    model: flags.model as string | undefined,
    temperature: num(flags.temperature),
    maxTokens: num(flags["max-tokens"]),
  };
}

File: packages/py/llmorch-cli/src/llmorch_cli/flags.py
"""Global-flag resolution — twin of packages/ts/cli/src/flags.ts.

Precedence: explicit flag > env var > (left to the request/config layer) default.
Only ``--provider`` has an env layer of its own (``LLMORCH_PROVIDER``).
"""

import os
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any


@dataclass
class ResolvedFlags:
    provider: str | None = None
    model: str | None = None
    temperature: float | None = None
    max_tokens: int | None = None


def _as_str(value: Any) -> str | None:
    return value if isinstance(value, str) else None


def resolve_flags(opts: Mapping[str, Any], env: Mapping[str, str] | None = None) -> ResolvedFlags:
    """Resolve the four global options with precedence flag > env."""
    env = env if env is not None else os.environ
    temperature = opts.get("temperature")
    max_tokens = opts.get("max_tokens")
    return ResolvedFlags(
        provider=_as_str(opts.get("provider")) or env.get("LLMORCH_PROVIDER"),
        model=_as_str(opts.get("model")),
        temperature=float(temperature) if isinstance(temperature, (int, float)) else None,
        max_tokens=int(max_tokens) if isinstance(max_tokens, int) else None,
    )
